2019/d07Amplifiers.d

```diff
@@ -0,0 +1,131 @@
+auto thrusterSignal(int[] program, int[] phases) {
+  import intcode;
+  int input;
+
+  foreach(phase; phases) {
+    input = IntCode(program.dup).run([phase, input])[0];
+  }
+
+  return input;
+}
+
+auto findMaxThrusterSignal(T)(int[] program, int phase, T callback) {
+  import std.algorithm, std.range, std.typecons;
+
+  int maxSignal;
+  int[] maxPhases;
+
+  foreach(phases; iota(phase, phase+5).permutations) {
+    int signal = callback(program, phases.array);
```

Comment:
The callback takes an `int[]` when all it needs is a forward range. There must be a way to specify this.